File: Hard/Alien Dictionary/alien-dictionary.py
from collections import deque , defaultdict
class Solution:
    def findOrder(self, words):
        # code here
        # code here
        mp = defaultdict(int)
        adj = defaultdict(list)
        indegree = defaultdict(int)
        
        for word in words:
            for ch in word:
                mp[ch] = 0
        
        for i in range(len(words) - 1):
            w1 = words[i]
            w2 = words[i+1]
            for j in range(len(w1)):
                
                if j>=len(w2):
                    return ""
                
                if w1[j] not in mp:
                    mp[w1[j]] = 0
                    
                if w2[j] not in mp:
                    mp[w2[j]] = 0
                
                if w1[j] != w2[j]:
                    adj[w1[j]].append(w2[j])
                    indegree[w2[j]] += 1
                    break

        # print(adj)
        
        q = deque()
        ans = []
        
        for i in mp:
            if i not in indegree:
                q.append(i)

        # print(indegree)
        # print(q)
        
        while(len(q) != 0):
            cur = q.popleft()
            ans.append(cur)
            
            for v in adj[cur]:
                indegree[v] -= 1
                if indegree[v] == 0:
                    q.append(v)
        
        if len(ans) != len(mp):
            return ""
        
        # print(ans)
        return ans
        

File: Hard/Alien Dictionary/alien-dictionary.cpp
class Solution {
	public:
	string findOrder(vector<string> &words) {
		// Store all unique characters present in the dictionary
		unordered_set<char> chars;
		
		for (string &s : words) {
			for (char &ch : s) {
				chars.insert(ch);
			}
		}
		
		// Total number of unique characters
		int V = chars.size();
		
		// Adjacency list representing precedence between characters
		unordered_map<char, string> adj(V);
		
		// Stores indegree of each character
		unordered_map<char, int> indegree;
		
		// Build the graph by comparing adjacent words
		for (int i = 0; i < words.size() - 1; i++) {
			string w1 = words[i];
			string w2 = words[i + 1];
			
			for (int j = 0; j < w1.size(); j++) {
				
				// Invalid case:
				// If the second word is a prefix of the first,
				// then no valid ordering exists
				if (j == w2.size())
					return "";
				
				// The first differing characters determine the ordering
				if (w1[j] != w2[j]) {
					adj[w1[j]].push_back(w2[j]);
					indegree[w2[j]]++;
					break;
				}
			}
		}
		
		// Initialize queue with all characters having zero indegree
		queue<char> q;
		
		for (char ch : chars) {
			if (indegree.count(ch) == 0) {
				q.push(ch);
			}
		}
		
		string ans = "";
		
		// Perform Kahn's Topological Sort
		while (!q.empty()) {
			char cur = q.front();
			q.pop();
			
			ans.push_back(cur);
			
			// Reduce indegree of all neighboring characters
			for (char v : adj[cur]) {
				indegree[v]--;
				
				// Push into queue when indegree becomes zero
				if (indegree[v] == 0)
					q.push(v);
			}
		}
		
		// If all characters are processed, return the ordering.
		// Otherwise, a cycle exists and no valid ordering is possible.
		return ans.size() == chars.size() ? ans : "";
	}
};
